modules/node_groups/node_groups.tf

```
@@ -0,0 +1,49 @@
+resource "aws_eks_node_group" "workers" {
+  for_each = random_pet.node_groups
```

A last question. So now we don't have intermediate locals anymore, can't we just loop over `local.node_groups_expanded` instead of `random_pet.node_groups` ? So we can use `each.value` (the current node group) in the ressource defintion ? I think it will be more readable, will avoid using split and different variable name for the same thing.

Furthermore, we only need the random pet for the `node_group_name`, so we can have something like this:

```hcl
node_group_name = join("-", [var.cluster_name, each.key, random_pet.node_groups[each.key].id])
```

@dpiddockcmp what do you think ?

---

I agree with @barryib. Terraform is less error-prone when you share a loop index between resources. In fact the simplest solution is to just use
```suggestion
  for_each = var.node_groups
```

---

I agree that the join and split is ... not great. But `random_pet.keepers` can only take a plain map(string). This is a limitation from terraform's provider API which does not support dynamic typing.

We need to pass in all those variables to the `random_pet` because they trigger a recreation on the `aws_eks_node_group` resource.

I then used random_pet.keepers in the aws_eks_node_group to reduce duplication of logic. For example on the instance_type:

```HCL
resource "random_pet" "node_groups" {
  for_each = local.node_groups_expanded
  keepers = {
    instance_type = lookup(each.value, "instance_type", var.workers_group_defaults["instance_type"])
  }
}

resource "aws_eks_node_group" "workers" {
  for_each = random_pet.node_groups
  instance_types = [each.value.keepers.instance_type]
}
```

Any change to the logic on random_pet is automatically received by the aws_eks_node_group. I did this as I made mistakes when migrating the old code which had duplication of logic.

We can go back to duplication of logic if you want but it will need careful review of any PRs in future that touch either of those resources 🤷‍♂ 

`labels` and `tags` refer to `var.node_groups_defaults` and `var.node_groups[each.key]` as terraform does not do a deep merge of maps.

---

If looping over `var.node_groups` you can still reference `random_pet.node_groups`, simply do the following replacement inside the resource block:
`each.value` => `random_pet.node_groups[each.key]`

Better yet, if you loop over local `local.node_groups_expanded` instead - which will make it more consistent with random_pet's index - you can do the following replacement as well:
`local.node_groups_expanded[each.key]` => `each.value`

---

There are currently 8 references to `each.value.keepers` and only 5 references to `local.node_groups_expanded`. Switching to the local for the `for_each` would make the code more verbose. 

The random provider also makes reference to "reading" variables through the keepers map in the example to ensure they change together: [docs](https://www.terraform.io/docs/providers/random/index.html#resource-quot-keepers-quot-)

---

I think we can resolve this conversation to let @max-rocket-internet ship a new release and get feedbacks from users. This conversation is more a coding style discussion and we can change it easily if needed in a next release.

---

Had a thought. I've moved the `var.workers_group_defaults` merging into the `local.node_groups_expanded` map. Allowed to greatly simplify the logic in `random_pet`. And thus no need to use it as a pass through.

---

Love it !